docs/source/development/code_testing.md

```diff
@@ -0,0 +1,76 @@
+---
+jupytext:
+  formats: md:myst
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.13.8
+kernelspec:
+  display_name: Python 3
+  language: python
+  name: pyrealm_python3
+---
+
+# Package testing and profiling
+
+The `pyrealm` package uses `pytest` to provide benchmark tests, unit tests and
+integration testing. In addition, `doctest` is used to maintain examples of code usage
+in the package docstrings and ensure that the documented return values are correct.
+
+## Using `pytest`
+
+The `tests` directory contains modules providing test suites for each of the different
+package modules. This includes:
+
+* regression testing the output of `pyrealm` code against previously existing
+  implementations of some functionality, such as the `rpmodel` and `SPLASH` packages.
+* unit testing of individual functions and methods and,
+* profiling and integration testing using combinations of modules.
+
+These are the main tests that ensure that the package is behaving as expected and that
+it produces stable outputs. The test suite can be run from repository using:
+
+```bash
+poetry run pytest
+```
+
+The `setup.cfg` file contains `pytest` configuration details. By default, we do not
+include the slow-running `profiling` tests as part of the standard `pytest` suite. See
+the [profiling page](./profiling_and_benchmarking.md) for details on running those
+tests.
+
+## Using `doctest`
+
+Some of the package docstrings contain `doctest` examples of code use. These examples
+are intended to provide simple examples of method or function use and generate an
+output: the `doctest` module is used to make sure that the code runs and gives the
+expected result.
+
+We have configured `pytest` to automatically also run `doctest`, but you can manually
+check the tests in files using, for example:
+
+```bash
+poetry run python -m doctest pyrealm/pmodel/pmodel.py
+```
+
+Normally, `doctest` is just used to test a return value: the value tested is the value
+printed to the console, so it is common to use some form of `round` to make sure values
+match. It can also be used to check that an error or warning is raised. See the
+docstring for {meth}`~pyrealm.core.hygro.convert_rh_to_vpd` to see how checking for
+warning text can be included in a doctest.
+
+## Using `pytest-coverage` and `codecov`
+
+Using the plugin [pytest-coverage](https://pypi.org/project/pytest-cov/) you can
+generate coverage reports. You can run:
+
+```bash
+poetry run pytest --cov=<test_path>
+```
+
+to perform coverage analysis. The report is stored with the name `index.html`. It can be
+used to determine if your contribution is adequately tested. The GitHub Actions
```

Comment:
Is this part still up-to-date after recent changes, @AmyOctoCat ?

Reply:
My feeling is that the auto-profiling is becoming more of a barrier to progress than an asset. There doesn't seem to be a good, minimally intrusive way of implementing this and we are getting profiling failures on PRs that don't affect the code. My guess is that this is some combination of the particular runner and the load being small enough to get frequent stochastic differences.

I honestly think we should dump the automatic profiling on PRs for now because:

* it is fouling up our PRs, 
* doesn't look feasible as we want it to run, and
* is unreliable.

We can re-use the code - maybe at another point? I think a manual process is probably easier - but we could try and generate a comparison (on the same runner) between a new release and its predecessor. That runs the risk of interfering with the trusted publishing though. Do any other projects do this kind of thing (successfully!).

Reply:
In the short term, I'm not inclined to hold up `1.0.0` while we iron this out - it's a developer facing issue that doesn't affect the release of the new functionality.